src/lib/homestore_backend/hs_pg_manager.cpp

```diff
@@ -48,32 +55,138 @@ PGError toPgError(ReplServiceError const& e) {
 }
 
 PGManager::NullAsyncResult HSHomeObject::_create_pg(PGInfo&& pg_info, std::set< peer_id_t > const& peers) {
+    auto pg_id = pg_info.id;
+    {
+        auto lg = std::shared_lock(_pg_lock);
+        auto it = _pg_map.find(pg_id);
+        if (_pg_map.end() != it) return folly::makeUnexpected(PGError::PG_ALREADY_EXISTS);
+    }
+
     pg_info.replica_set_uuid = boost::uuids::random_generator()();
     return hs_repl_service()
         .create_repl_dev(pg_info.replica_set_uuid, peers)
         .via(executor_)
-        .thenValue([this, pg_info = std::move(pg_info)](auto&& v) mutable -> PGManager::NullResult {
+        .thenValue([this, pg_info = std::move(pg_info)](auto&& v) mutable -> PGManager::NullAsyncResult {
             if (v.hasError()) { return folly::makeUnexpected(toPgError(v.error())); }
+            // we will write a PGHeader to the raft group and commit it when PGHeader is commited on raft
+            // group. so that all raft members will create the PGinfo and index table for this PG.
+            return do_create_pg(v.value(), std::move(pg_info));
+        });
+}
+
+PGManager::NullAsyncResult HSHomeObject::do_create_pg(cshared< homestore::ReplDev > repl_dev, PGInfo&& pg_info) {
+    auto serailized_pg_info = serialize_pg_info(pg_info);
+    auto msg_size = sisl::round_up(serailized_pg_info.size(), repl_dev->get_blk_size());
+    auto req = repl_result_ctx< PGManager::NullResult >::make(msg_size, 512 /*alignment*/);
+
+    auto buf_ptr = req->hdr_buf_.bytes();
+    std::memset(buf_ptr, 0, msg_size);
+    std::memcpy(buf_ptr, serailized_pg_info.c_str(), serailized_pg_info.size());
+
+    req->header_.msg_type = ReplicationMessageType::CREATE_PG_MSG;
+    req->header_.pg_id = pg_info.id;
+    req->header_.shard_id = 0;
+    req->header_.payload_size = msg_size;
+    req->header_.payload_crc = crc32_ieee(init_crc32, buf_ptr, msg_size);
+    req->header_.seal();
+    sisl::blob header;
+    header.set_bytes(r_cast< uint8_t* >(&req->header_));
+    header.set_size(sizeof(req->header_));
```

Comment:
I know this may be the pattern so far; but I would suggest writing some free (non-member) functions in this source that just do type conversion for building requests and their headers; that way we can write unit tests around them without having to exercise the raft code.

Reply:
will creat separate PR to move this part to a free function and add the corresponding UT